File: pdui_flutter/lib/core/pdui_debug_print.dart
import 'package:flutter/material.dart';

import '../models/pdui_fn.dart';
import '../proto_out/pdui-proto-out.pb.dart';

class PduiDebugPrint extends PduiFn {
  PBDebugPrint pbDebugPrint;

  PduiDebugPrint(this.pbDebugPrint) : super("debugPrint");

  @override
  resolve() {
    late String debugPrintBody;

    //TODO: Strategy?
    switch (pbDebugPrint.expression.identity) {
      case "scaffold":
        debugPrintBody = pbDebugPrint.expression.scaffold.toString();
        break;
      case "center":
        debugPrintBody = pbDebugPrint.expression.center.toString();
        break;
      case "column":
        debugPrintBody = pbDebugPrint.expression.column.toString();
        break;
      case "text":
        debugPrintBody = pbDebugPrint.expression.elevatedButton.toString();
        break;
      case "elevatedButton":
        debugPrintBody = pbDebugPrint.expression.elevatedButton.toString();
        break;
      case "string":
        debugPrintBody = pbDebugPrint.expression.rawString;
        break;
      default:
        debugPrintBody = "[ERROR] Payload not found by debugPrint";
    }

    debugPrint(debugPrintBody);
  }
}

File: pdui-express/lib/core/PDUIDebugPrint.ts
import { PDUIFn } from "../models/PDUIFn";
import { PDUIWidget } from "../models/PDUIWidget";
import {
    PBFnExpression,
    type PBDebugPrint,
    type PBExpression,
} from "../proto-out/pdui-proto-out.ts";

export class PDUIDebugPrint extends PDUIFn implements PBDebugPrint {
    expression: PBExpression;

    constructor(expression: any) {
        super({
            identity: "debugPrint",
            name: "debugPrint",
            params: {},
            body: {},
        });
        switch (typeof expression) {
            case "number":
            case "string":
                this.expression = {
                    identity: "string",
                    payload: {
                        oneofKind: "rawString",
                        rawString: expression.toString(),
                    },
                };
                break;

            case "object":
                const widget = expression as PDUIWidget;
                this.expression = widget.toExpression();
                break;
            default:
                throw Error("Wrong type used on PDUIDebugPrint");
        }
    }

    override toPB(): PBDebugPrint {
        return { expression: this.expression };
    }

    override toFnExpression(): PBFnExpression {
        return {
            identity: "debugPrint",
            payload: {
                oneofKind: "debugPrint",
                debugPrint: {
                    expression: {
                        identity: this.expression?.identity,
                        key: this.expression?.key,
                        payload: this.expression.payload,
                    },
                },
            },
        };
    }
}
